wire-runtime/src/test/proto-java/com/squareup/wire/map/Mappy.java

```diff
@@ -0,0 +1,140 @@
+
+  private static final long serialVersionUID = 0L;
+
+  @WireField(
+      tag = 1,
+      keyAdapter = "com.squareup.wire.ProtoAdapter#STRING",
+      adapter = "com.squareup.wire.map.Thing#ADAPTER"
+  )
+  public final Map<String, Thing> things;
+
+  public Mappy(Map<String, Thing> things) {
+    this(things, ByteString.EMPTY);
+  }
+
+  public Mappy(Map<String, Thing> things, ByteString unknownFields) {
+    super(ADAPTER, unknownFields);
+    this.things = Internal.immutableCopyOf("things", things);
+  }
+
+  @Override
+  public Builder newBuilder() {
+    Builder builder = new Builder();
+    builder.things = Internal.copyOf("things", things);
+    builder.addUnknownFields(unknownFields());
+    return builder;
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    if (other == this) return true;
+    if (!(other instanceof Mappy)) return false;
+    Mappy o = (Mappy) other;
+    return unknownFields().equals(o.unknownFields())
+        && Internal.equals(things, o.things);
+  }
+
+  @Override
+  public int hashCode() {
+    int result = super.hashCode;
+    if (result == 0) {
+      result = unknownFields().hashCode();
+      result = result * 37 + (things != null ? things.hashCode() : 0);
+      super.hashCode = result;
+    }
+    return result;
+  }
+
+  @Override
+  public String toString() {
+    StringBuilder builder = new StringBuilder();
+    if (!things.isEmpty()) builder.append(", things=").append(things);
+    return builder.replace(0, 2, "Mappy{").append('}').toString();
+  }
+
+  public static final class Builder extends Message.Builder<Mappy, Builder> {
+    public Map<String, Thing> things;
+
+    public Builder() {
+      things = Internal.newMutableMap();
+    }
+
+    public Builder things(Map<String, Thing> things) {
+      Internal.checkElementsNotNull(things);
+      this.things = things;
+      return this;
+    }
+
+    @Override
+    public Mappy build() {
+      return new Mappy(things, super.buildUnknownFields());
+    }
+  }
+
+  private static final class ProtoAdapter_Mappy extends ProtoAdapter<Mappy> {
+    private final ProtoAdapter<Map<String, Thing>> things = ProtoAdapter.createMapAdapter(ProtoAdapter.STRING, Thing.ADAPTER);
+
+    ProtoAdapter_Mappy() {
+      super(FieldEncoding.LENGTH_DELIMITED, Mappy.class);
+    }
+
+    @Override
+    public int encodedSize(Mappy value) {
+      return things.encodedSizeWithTag(1, value.things)
+          + value.unknownFields().size();
+    }
+
+    @Override
+    public void encode(ProtoWriter writer, Mappy value) throws IOException {
+      things.encodeWithTag(writer, 1, value.things);
+      writer.writeBytes(value.unknownFields());
+    }
+
+    @Override
+    public Mappy decode(ProtoReader reader) throws IOException {
+      Builder builder = new Builder();
+      long token = reader.beginMessage();
+      for (int tag; (tag = reader.nextTag()) != -1;) {
+        switch (tag) {
+          case 1: builder.things.putAll(things.decode(reader)); break;
```

Comment:
... and this is the code that decodes a single entry. Still tempted to make `things` a `ProtoAdapter<Map.Entry<String, Thing>>`